preprocess/docenc_main.cc

```diff
@@ -0,0 +1,258 @@
+#include <vector>
+#include "util/file_stream.hh"
+#include "util/file_piece.hh"
+#include "util/string_piece.hh"
+#include "preprocess/base64.hh"
+
+using namespace std;
+
+enum Mode {
+	COMPRESS,
+	DECOMPRESS
+};
+
+string prefix_lines(string const &input, string const prefix)
+{
+	stringstream sin(input);
+	stringstream sout;
+
+	string line;
+	while (getline(sin, line))
+		sout << prefix << line << '\n';
+
+	return sout.str();
+}
+
+size_t decode(util::FilePiece &in, util::FileStream &out, char delimiter, vector<size_t> const &indices, bool print_document_index, bool &delimiter_encountered) {
+	size_t document_index = 0;
+	vector<size_t>::const_iterator indices_it(indices.begin());
+
+
+	for (StringPiece line : in) {
+		++document_index;
+
+		if (!indices.empty()) {
+			if (*indices_it != document_index) {
+				continue; // skip document
+			} else {
+				indices_it++;
+			}
+		}
+
+		string document;
+		preprocess::base64_decode(line, document);
+
+		if (!delimiter_encountered && document.find(delimiter == '\n' ? string("\n\n") : string(&delimiter, 1)) != string::npos)
+			delimiter_encountered = true;
+
+		if (print_document_index)
+			document = prefix_lines(document, to_string(document_index) + "\t");
+
+		out << document << delimiter;
+
+		// Have we found all our indices? Then stop early
+		if (!indices.empty() && indices_it == indices.end())
+			break;
+	}
+
+	return document_index;
+}
+
+size_t encode(util::FilePiece &in, util::FileStream &out, char delimiter, vector<size_t> const &indices) {
+	size_t document_index = 0;
+	string document;
+	vector<size_t>::const_iterator indices_it(indices.begin());
+
+	bool is_eof = false;
+	while (!is_eof) {
+		document.clear();
+		
+		// Start accumulating lines that make up a document
+		StringPiece line;
+		while (true) {
+			is_eof = !in.ReadLineOrEOF(line, delimiter, true);
+			
+			if (is_eof)
+				break;
+
+			// Is this the document delimiter when using \n\n as delimiter?
+			if (delimiter == '\n' && line.empty())
+				break;
+			
+			document.append(line.data(), line.size());
+			
+			// Add back the \n delimiter for lines
+			if (delimiter == '\n')
+				document.push_back('\n');
+
+			if (delimiter != '\n')
+				break;
+		}
+
+		// Don't bother printing anything for that last empty doc
+		if (is_eof && document.empty())
+			break;
+		
+		++document_index;
+
+		// Check whether this is a document we care about
+		if (!indices.empty()) {
+			if (*indices_it != document_index) {
+				continue; // skip document
+			} else {
+				indices_it++;
+
+				// Check whether we can stop processing altogether after this one 
+				if (indices_it == indices.end())
+					is_eof = true;
+			}
+		}
+
+		string encoded_document;
+		preprocess::base64_encode(StringPiece(document.data(), document.size()), encoded_document);
+		out << encoded_document << '\n';
+	}
+
+	return document_index;
+}
+
+int usage(char program_name[]) {
+	cerr << "Usage: "
+	     << program_name << " [ index ... ] [ files ... ]\n"
```

Comment:
What does this program do?

Reply:
I've added a few examples to the readme, but in the pipeline I'm only using it in europat, to convert from Omniscien's tabular format where lines are encoded using HTML tags to base64, without writing a for-loop in bash. E.g:
```
cat patents.tab \
    | cut -f3 \
    | sed -r 's/$/\x0/g' \
    | sed -r 's/<br\/>|<\/p><p>/\n/g' \
    | sed -r 's/<\/?p>//g' \
    | docenc -0
```

The prefix encoding is never used in the pipeline, it's just there for me so I can do `docenc -d -n sentences.gz | fgrep 'weird stuff'` and find which document it was exactly.